### ISSUE TITLE
[BUG] windows paths provided to coverage map

### ISSUE THREAD

**Issue by piotr-cz** [tapjs/tapjs#966]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Have you read the `CONTRIBUTING` guide on posting bugs, and `CODE_OF_CONDUCT`?

- [X] yes I read the things

### This issue exists in the latest tap version

- [X] I am using the latest tap

### Description

Coverage reports are not being written to disk

### Reproduction

- install any version from 18.4.3 to 18.5.3
  - execute `npx tap --coverage-report=cobertura`
  - coverage report **is not written** to `.tap/report/cobertura-coverage.xml`
- install v18.4.2
  - execute `npx tap --coverage-report=cobertura`
  - coverage report **is written** to `.tap/report/cobertura-coverage.xml`

### Environment

```
npm ls tap
tap@18.5.3

tap versions
tap: 18.5.3
"@tapjs/config": 2.4.9
"@tapjs/core": 1.4.2
"@tapjs/run": 1.4.10
"@tapjs/stack": 1.2.6
"@tapjs/test": 1.3.13
tap-parser: 15.3.0
tap-yaml: 2.2.0
tcompare: 6.4.3
plugins:
  "@tapjs/after": 1.1.13
  "@tapjs/after-each": 1.1.13
  "@tapjs/asserts": 1.1.13
  "@tapjs/before": 1.1.13
  "@tapjs/before-each": 1.1.13
  "@tapjs/filter": 1.2.13
  "@tapjs/fixture": 1.2.13
  "@tapjs/intercept": 1.2.13
  "@tapjs/mock": 1.2.11
  "@tapjs/node-serialize": 1.2.2
  "@tapjs/snapshot": 1.2.13
  "@tapjs/spawn": 1.1.13
  "@tapjs/stdin": 1.1.13
  "@tapjs/typescript": 1.3.2
  "@tapjs/worker": 1.1.13

tap config list
# vim: set filetype=yaml :

# from .taprc
allow-incomplete-coverage: true
coverage-map: coverage-map.js
include:
  - test/**/*.test.ts
test-env:
  - NODE_ENV=test
timeout: 45
typecheck: false

# env, cli, and defaults
color: true
coverage-report:
  - text
exclude:
  - "**/@(fixture*(s)|dist)/**"
jobs: 6
reporter: base
snapshot-clean-cwd: true

tap plugin list
@tapjs/after
@tapjs/after-each
@tapjs/asserts
@tapjs/before
@tapjs/before-each
@tapjs/filter
@tapjs/fixture
@tapjs/intercept
@tapjs/mock
@tapjs/node-serialize
@tapjs/snapshot
@tapjs/spawn
@tapjs/stdin
@tapjs/typescript
@tapjs/worker

# if using typescript:
npm ls ts-node
ts-node@10.9.1

npm ls typescript
typescript@5.2.2

# posix:
-

# windows (in powershell)
Microsoft Windows NT 10.0.19045.0 x64
```

**Comment by piotr-cz** [comment]
btw 
- the bug probably was introduced somewhere here: https://github.com/tapjs/tapjs/compare/tap%4018.4.2...tap%4018.4.3
- in affected versions the `.tap/report` directory is not being created (which prevents generating coverage)

**Comment by isaacs** [comment]
Hmm, I can't seem to reproduce this on my system with any version of tap, and it's pretty surprising that it would have changed between 18.4.2 and 18.5.3, since nothing in that area was touched.

I can see the report generator in `istanbul-lib-reports` is at least attempting to create the directory before attempting to write the file. https://github.com/istanbuljs/istanbuljs/blob/master/packages/istanbul-lib-report/lib/file-writer.js#L184

It does appear that istanbul-lib-report is using make-dir rather than mkdirp, maybe it's failing somehow on Windows? I don't know why it would, though.

What does `ls -laF .tap/report` show? Is there a file in the way or something? Maybe tap is swallowing an error that it shouldn't be?

**Comment by piotr-cz** [comment]
**versions >= 18.4.3**

At the end, there are no `report` or `coverage` directories generated, only

```
processinfo
test-results
```

Command
```
npx tap --coverage-report=cobertura
```

Output
```

  🌈 TEST COMPLETE 🌈


Asserts:  168 pass  0 fail  168 of 168 complete
Suites:    14 pass  0 fail    14 of 14 complete

# No coverage generated
# { total: 168, pass: 168 }
# time=43990.333ms
```

**version 18.4.3**

at the end following directories within `.tap` directory are generated:
```
coverage
processinfo
report
test-results
```

Command
```
npx tap --coverage-report=cobertura
```

Output
```
cobertura report: xxx\.tap\report\cobertura-coverage.xml


  🌈 TEST COMPLETE 🌈


Asserts:  182 pass  0 fail  182 of 182 complete
Suites:    14 pass  0 fail    14 of 14 complete

# ERROR: incomplete statements coverage (91.36%)
# ERROR: incomplete branches coverage (68.88%)
# ERROR: incomplete functions coverage (87.38%)
# ERROR: incomplete lines coverage (91.36%)
# { total: 182, pass: 182 }
# time=62797.752ms
```

**Comment by piotr-cz** [comment]
I traced the issue to this change: [src/run/src/run.ts](https://github.com/tapjs/tapjs/compare/tap@18.4.2...tap@18.4.3#diff-d4faa1c1242db699819fec45f2277a8aad63c655cb6855e9bf15173e3da7c5a9)

```diff
+      if (mapped?.length && !coveredFiles?.length) {
+        // provided a glob expression that didn't match anything.
+        // cover nothing.
+        coveredFiles = null
+      }
```

Indeed this might by windows - specific

**Comment by piotr-cz** [comment]
I fixed the issue by tweaking my [coverage map](https://node-tap.org/coverage/#coverage-maps):

```diff
 const regExp = /^test[\\/](.*)\.test\.ts/

 export default test =>
   regExp.test(test)
-    ? test.replace(regExp, 'src/$1.js')
+    ? test
+        .replace(regExp, 'src/$1.ts')
+        .replace(/\\/g, '/')
     : null
```

**Comment by isaacs** [comment]
Ahh, I see, yeah, that is indeed a bug. It should be only providing posix-style path separators consistently to the coverage-map function, which would mean that you don't have to do that extra step.

**Comment by isaacs** [comment]
Hm, it looks like that was already fixed on 18.5.3, with commit 55503012.

Can you confirm that you're getting `\` in the paths provided to the coverage map in that version?

**Comment by piotr-cz** [comment]
I can confirm that this has been fixed in 18.5.3.

I had just to fix my coverage map like that:

```diff
 const regExp = /^test[\\/](.*)\.test\.ts/

 export default test =>
   regExp.test(test)
-    ? test.replace(regExp, 'src/$1.js')
+    ? test.replace(regExp, 'src/$1.ts')
     : null
```

Thank you for assistance!

**Comment by isaacs** [comment]
Ah! Right, because `src` has `.ts` files, not `.js` files.

I stared at that for way longer than I'd like to admit before I saw what the diff even was 😅
